docs/BRANCHING.md

```diff
@@ -0,0 +1,51 @@
+
+## General Considerations
+
+- Generally the master branch is the one where current development is happening.
+- One objective would be to arrive to a point where the master branch can be released very often (every day?).
+  This means the master branch is releasable at all times. I believe this strategy is called **Continuous Delivery**.
+- At this point we don't think it's possible to do that until the testing infrastructure has been improved.
+- Nevertheless, the branching strategy takes this objective into account.
+- To further support this objective, the strategy wil try to minimize the number of long-lasting branches.
+
+## Release numbering
+- example 4.3.0
+- The first number (4 in this example) is incremented for major changes in the software. What major change mean has to be decided.
+- The second number (3 in this example) is incremented for minor changes in the software.
+- The third number (0 in this example) is used for maintenance releases.
+  A maintenance is created if some bug has to be corrected before the next minor release.
+![](BranchingDiagram.png)
+## Feature/fix branches
+
+- Feature/fix branch are created for every feature we want to add to the software (or bug we want to fix).
+- We will call them feature branches for the rest of this document.
+- They stem from master, are usually short-lived and are merged back to master.
+
+## Maintenance/release branches
+- In general we will try not to create long-lived release branches.
+- A long-lived release branch will be created only if any merge to master that came after the previous release is deemed not suitable to go in the maintenance release.
+- For example the change could be too extensive, or introduce an undue risk in the maintenance release.
+  This is true in particular if the testing infrastructure is not sufficient.
+
+## Hotfixes
+- A hotfix will generally result in a maintenance release.
+- Hotfixes branches are similar to feature branches.
+- Where does the hotfix branch stem from:
+    - Generally from the commit that was tagged for the latest release.
+        - See for example commit A in the diagram.
+    - In keeping with the **Continuous delivery** objective, we will allow the hotfix branch to stem from a commit
+      downstream of the previous release commit if:
+        - we consider that all merges coming from feature branches that happened after the previous release can go in the maintenance release.
+            - For example if commit C from the diagram contains minor changes like typos we could decide that it's OK to include it in the maintenance release.
+            - In that case the hotfix branch can stem from commit C and be merged back into master in commit F.
+            - Commit F is tagged as the maintenance release and the maintenance release is built from there.
+    - If a merge on the master branch cannot be included in the maintenance release (See for example commit I) then we have no choice but to create a maintenance branch
```

Comment:
@jcpitre So basically, if there's been a commit to master that would need to result in a major release version, then we have to create a maintenance branch?

Reply:
If there is something we absolutely need to fix in the released software, yes.
Considering that normally you would not want the major commit to make it's way in the maintenance release.
But this can be true for minor commits too. Except it's not as clear cut.